### PR TITLE
Remove and ban ImmutableMap

### PR DESCRIPTION
**Pull request by nik9000**
Removes and bans `ImmutableMap`. This is one of the last blockers for #13224.

### CODE REVIEW

**Comment by nik9000** [comment]
I'm trying to do this in a minimally invasive way - no switching to `Streams` API.

I've added a forbidden-apis check to stop us from using `unmodifiableMap` in the `org.elasticsearch.cluster` package, with the intent of forcing us into using ImmutableOpenMap that is similarly compact.

**Comment by nik9000** [comment]
I should not that there are two kinds of concerns around using `unmodifiableMap` in the `org.elasticsearch.cluster` package:
1. Space: `unmodifiableMap` wrapping a `HashMap` presumably takes more space than `ImmutableMap`. `ImmutableMap` uses open addressing and `HashMap` uses chains and/or balanced trees to resolve conflicts.
2. Heap churn: cluster state is immutable - meaning that an object reference into cluster state will always reference the same cluster state. To change the cluster state build copies of it. This involves a lot of copying, not copying the entire thing, but still lots of copying. Larger data structures like `HashMap` involve more allocations which in turn will cause more stop the world pauses even in a healthy environment.

In this PR I aim to keep cluster state's maps in ones that use open addressing to keep them small-ish. So to not make number 1 above worse.

It would be interesting to investigate [persistent data structures](https://en.wikipedia.org/wiki/Persistent_data_structure) like `CopyOnWriteHashMap` which is a [hash array mapped trie](http://en.wikipedia.org/wiki/Hash_array_mapped_trie). But that should await another pull request.

**Comment by nik9000** [comment]
@jasontedor ready for review! Tons of little changes.

**Comment by jasontedor** [comment]
@nik9000 Any chance that you could rebase on master? There are a lot of unrelated changes because of the merge of master into this branch.

**Comment by nik9000** [comment]
> @nik9000 Any chance that you could rebase on master? There are a lot of unrelated changes because of the merge of master into this branch.

I can certainly merge master into it again if that'll help. Rebasing on master would be harder I think and I thought that wasn't the recommended thing anyway when you have large-ish branches like this.

**Comment by jasontedor** [comment]
@nik9000 Merging master in won't help, it will just add more noise to this diff. It's rebasing that will eliminate the noise because then the diff between what you merged in and this branch will be zero except for the changes that you made.

You'll have to rebase on master at some point anyway. :)

That said, if you think that rebasing on master will slow down the review process, then can you revert the commits ba68a8df63f45af12642b0874ac64dd4841fc832 and bb2611d2f55f8a07e6a33cb1c40059eed50f33c1?

At the end of the day, my goal is to remove the noise that isn't from your change, but is instead from changes that took place in master.

**Comment by nik9000** [comment]
> @nik9000 Merging master in won't help, it will just add more noise to this diff.

I talked with @jasontedor over video and chat and we came to the conclusion that github is broken for pull requests of this size. Really the only way to handle it is to pull the patch locally and diff it against master. We just don't trust github for PRs this large. We'll need to avoid pull requests of this size in future. Luckily they are reasonably rare for single features.

The talk about merging master into the pull request was a red herring - the pull request was really just that big.

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java [hunk 1]
```diff
@@ -514,12 +517,17 @@ public Builder version(long version) {
             return this;
         }
 
+        /**
+         * Builds the routing table. Note that this can only be called one time.
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: This should be enforced then (set a flag, check the flag, etc.).

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java]: I can add something more explicit but calling it more than once will throw an exception in the ImmutableOpenMap. It'd be nice to catch the exception and transform it into something more obvious.

File: core/src/main/java/org/elasticsearch/common/util/CollectionUtils.java [hunk 2]
```diff
@@ -458,4 +474,9 @@ public static int sortAndDedup(final BytesRefArray bytes, final int[] indices) {
         return result;
     }
 
+    public static <E> void addAll(Collection<E> collection, Iterable<ObjectCursor<E>> iterable) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/util/CollectionUtils.java]: This method is used in exactly one place. Can we just inline it there?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/util/CollectionUtils.java]: Certainly. I added it thinking "I'm sure I'll need this a few times" and then forgot about it.

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java [hunk 3]
```diff
@@ -208,7 +206,6 @@ public void testCanRemainUsesLeastAvailableSpace() {
         ShardRouting test_2 = ShardRouting.newUnassigned("test", 2, null, true, new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, "foo"));
         ShardRoutingHelper.initialize(test_2, node_1.getId());
         ShardRoutingHelper.moveToStarted(test_2);
-        shardRoutingMap.put(test_2, "/node1/most");
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java]: Is the removal (without replacement) of this intentional?

**Comment by jasontedor** [comment]
@nik9000 This is great work, and very much appreciated. I left a few comments but otherwise LGTM. Let's get this home.

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java (hunk 3 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java]: Yes but I'll make it more clear.

**Comment by nik9000** [comment]
> @nik9000 This is great work, and very much appreciated. I left a few comments but otherwise LGTM. Let's get this home.

I believe I've addressed you points. Have a look at  bfb9054, 56318df, and 9492223.

I'm going to merge master into this again to see if I can get a clean merge back to master when it comes time to merge it.

**Comment by jasontedor** [comment]
Yes, looks great. Just one more thing. I think the forbidden-api changes were helpful for development, but let's take them out before taking this PR to master?

**Comment by nik9000** [comment]
> Yes, looks great. Just one more thing. I think the forbidden-api changes were helpful for development, but let's take them out before taking this PR to master?

Done in beeea14.

**Comment by jasontedor** [comment]
LGTM.

**Comment by nik9000** [comment]
> LGTM.

Cool. I'll rerun test yet another time (the merge didn't go as well as I thought) and if they pass I'll merge.

**Pull request merged**